velox/docs/develop/memory.rst

```diff
@@ -0,0 +1,890 @@
+memory allocator in Velox handles all the memory allocations from both file
+cache and query memory. It ensures the total allocated memory won’t exceed the
+system memory limit configured for Velox to use. To further handle the spiky
+memory usage from the non-Velox components, Velox provides a generic server
+memory pushback mechanism to automatically shrink the file cache to return the
+unused Velox memory back to the OS when the server is detected under low memory
+condition.
+
+Overall Design
+--------------
+
+Velox memory system consists of the following major memory components: memory
+manager (*MemoryManager*), memory allocator (*MemoryAllocator*), file cache
+(*AsyncDataCache*), memory arbitrator (*MemoryArbitrator*) and a number of memory
+pools (*MemoryPool*). The memory manager creates all the other components. It
+creates the memory allocator and memory arbitrator when initializing the memory
+system, and creates the memory pool on-demand for query execution.
+
+.. image:: images/memory-system.png
+    :width: 600
+    :align: center
+
+When a query starts execution, it first creates a root memory pool (query pool)
+from the memory manager, and then creates a tree of child memory pools from the
+root according to the query plan: a child memory pool for each query task
+(task pool), a grandchild memory pool for each plan node (node pool) and a
+great-grandchild memory pool for each operator instance (operator pool). During
+the query execution, it allocates memory from the leaf operator pools and
+propagates the memory usage up to the root query pool. If the aggregated memory
+usage at the root exceeds the current query memory capacity, the query pool
+sends a request to the memory arbitrator to grow its capacity. The memory
+arbitrator either grows the requestor query pool’s capacity by reclaiming the
+used memory from the other queries which has the largest capacity in the
+system, or reclaiming from the request query pool itself to free up space
+within its current capacity if it exceeds the per-query memory limit or it has
+the largest capacity in the system. The used memory reclaim is achieved through
+techniques like `disk spilling <https://facebookincubator.github.io/velox/develop/spilling.html>`_. If the memory arbitration succeeds, the leaf
+operator pool can proceed with the actual memory allocation from memory
+allocator. However, if the memory arbitration fails, then the query which has
+the largest capacity in the system is chosen to fail with query memory capacity
+exceeded error (local OOM). The failed query may or may not be the requestor
+query pool itself.
+
+The memory allocator does the actual memory allocation from its own managed
+memory space in units of machine pages (4KB). It tracks the amount of allocated
+memory, and returns an error if an allocation request exceeds the system memory
+limit. This enables explicit control on RSS of Velox memory usage to help
```

Comment:
s/RSS/RSS (Resident Set Size) ? I used to think RSS was a Remote Shuffle Service :)

Reply:
Resident Set Size is abbreviated as RSS at the first place in this doc? So we assume RSS is Resident Set Size? Thanks!